File: src/main/java/ar/com/flow/bankaccount/Withdrawal.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.transaction.Step;
import ar.com.flow.bankaccount.transaction.Transaction;
import ar.com.flow.bankaccount.transaction.receipt.Action;
import ar.com.flow.bankaccount.transaction.receipt.DebitPrinter;
import ar.com.flow.bankaccount.transaction.receipt.ReceiptPrinter;
import ar.com.flow.bankaccount.withdrawal.SufficientFunds;
import ar.com.flow.money.Money;

class Withdrawal {
    static WithdrawalBuilder from(BankAccount debitAccount) {
        return new WithdrawalBuilder(debitAccount);
    }

    static class WithdrawalBuilder {
        private final BankAccount debitAccount;

        public WithdrawalBuilder(BankAccount debitAccount) {
            this.debitAccount = debitAccount;
        }

        public Transaction amount(Money amountToWithdraw) {
            return Transaction.builder()
                    .amount(amountToWithdraw)
                    .precondition(new SufficientFunds(debitAccount, amountToWithdraw))
                    .step(new Step(debitAccount, Withdrawal.receipt(debitAccount)))
                    .build();
        }
    }

    static ReceiptPrinter receipt(BankAccount account) {
        return new DebitPrinter(account, Action.Withdrawal);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/Withdrawal.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.transaction.Builder
import ar.com.flow.bankaccount.transaction.steps.Step
import ar.com.flow.bankaccount.transaction.Transaction
import ar.com.flow.bankaccount.transaction.receipt.*
import ar.com.flow.bankaccount.withdrawal.SufficientFunds
import ar.com.flow.money.Money

internal object Withdrawal {
    fun from(debitAccount: BankAccount): WithdrawalBuilder {
        return WithdrawalBuilder(debitAccount)
    }

    fun receipt(account: BankAccount): ReceiptPrint {
        return DebitPrint(account, Action.Withdrawal)
    }

    fun scratch(account: BankAccount): ReceiptScratch {
        return DebitScratch(account, Action.Withdrawal)
    }

    internal class WithdrawalBuilder(private val debitAccount: BankAccount) {
        fun amount(amountToWithdraw: Money): Transaction {
            return Builder()
                .amount(amountToWithdraw)
                .precondition(
                    SufficientFunds(
                        debitAccount,
                        amountToWithdraw
                    )
                )
                .step(Step(debitAccount, receipt(debitAccount), scratch(debitAccount)))
                .build()
        }
    }
}